ci: move devnet test to make target

jyellick opened the pull request:
When encountering devnet test failures in CI, they're challenging to resolve and reproduce locally because the commands are embedded into the CI configuration and not the Makefile.

This PR simply extracts the commands from the CI steps and embeds them into the devnet python module, allowing for the python script to be invoked with a new `--test` flag to execute those steps.  It also adds a `devnet-test` make target to allow simple invocation of the script in test mode.

jyellick committed:
ci: move devnet test to make target

When encountering devnet test failures in CI, they're challenging to
resolve and reproduce locally because the commands are embedded into the
CI configuration and not the Makefile.

tynes commented on bedrock-devnet/devnet/__init__.py:
```diff
@@ -43,13 +46,20 @@ def main():
 
+    if args.test:
+      log.info('Testing deployed devnet')
+      devnet_test(paths)
```
The idea here is that this test command is independent of bringing up the stack?

jyellick replied:
Correct.  To run the test manually, you'd do something like:

```
make devnet-up-deploy
make devnet-test
```

I'm not sure how everyone is using the `devnet` targets, but it seems plausible that some users are running other loads against the devnet, so I was wary of including the test execution with bringing the devnet up or down.

trianglesphere replied:
you would have `devnet-up-deploy` be a pre-requisite for `devnet-test` if you wanted `devnet-test` to automatically bring it up. The only potential issue is that it might always restart the devnet.

jyellick replied:
Right.  Conceivably, we could have multiple targets if we wanted to allow both, say `devnet-test-only` and `devnet-test` which required the `devnet-up-deploy` followed by `devnet-test-only`.  If there's some consensus, I'm happy to modify.

tynes commented on .circleci/config.yml:
```diff
@@ -924,17 +924,9 @@ jobs:
-                name: Deposit ETH through the bridge
-                command: timeout 8m npx hardhat deposit-eth --network devnetL1 --l1-contracts-json-path ../../.devnet/sdk-addresses.json
-                working_directory: packages/sdk
+                name: Test the stack
```
This doesn't look like it's running in CI, not sure if it's a permissions issue

jyellick replied:
I'm certainly not a circle CI expert, but it did execute in the source repo https://app.circleci.com/pipelines/github/bobanetwork/v3-anchorage/97/workflows/8f93a8dd-291f-4452-ad14-c47411905136/jobs/6089 so I believe the definition should be alright.

tynes reviewed (commented):
Generally in favor of the idea, not sure how I feel about the approach but could be open to it

jyellick commented:
> Generally in favor of the idea, not sure how I feel about the approach but could be open to it

I'm certainly happy to rework in some other way.  As best as I could tell, the assorted build scripts are being migrated to python (vs. bash), so I tried to keep everything in the python, even though it's simply executing commands.

If there's a different preferred approach, please let me know.  My goal here is primarily to allow developers to iterate while running tests without having to push it out and wait on CI.

tynes commented:
> > Generally in favor of the idea, not sure how I feel about the approach but could be open to it
> 
> I'm certainly happy to rework in some other way. As best as I could tell, the assorted build scripts are being migrated to python (vs. bash), so I tried to keep everything in the python, even though it's simply executing commands.
> 
> If there's a different preferred approach, please let me know. My goal here is primarily to allow developers to iterate while running tests without having to push it out and wait on CI.

I am generally on board with this change after thinking about it more

ajsutton reviewed (commented):
Love this change - moving stuff from `.circleci/config.yml` into makefiles is a long term goal.  Ideally we'd also move the `devnet (with genesis contracts)` tests as well, though that could be a separate PR as this is definitely an improvement.

If we do move the genesis tests, it would be great to remove the duplication as well so `make devnet-test` could just work for both genesis and deploy modes. I think the only difference is if `--l1-contracts-json-path` is specified or not so possibly it could switch based on whether that file exists or not.

jyellick commented:
I've updated the `--test` path of the python to add or omit the sdk address path depending on whether the file exists.  Now, things should work both for `make devnet-up` and `make devnet-up-deploy` -- I've also updated the circle CI definition to use it in both paths.

jyellick commented:
Is there anything preventing this from merging?  Happy to make any desired changes if so.

jyellick commented:
Rebased to try to keep this current (though no conflicts existed).  Any updates from the maintainers? @tynes @ajsutton ?

ajsutton commented on .circleci/config.yml:
```diff
@@ -989,22 +989,8 @@ jobs:
-          name: Dump op-node logs
-          command: |
-            docker logs ops-bedrock-op-node-1 || echo "No logs."
-          when: on_fail
```
This piece doesn't seem to be in the python script so should probably stay here.

jyellick replied:
Oops! Good catch.  I hadn't originally removed this, but looks like it got clobbered in one of the rebases.  Fixed.

ajsutton reviewed (approved):
LGTM, other than I think we want to keep the capturing of op-node logs. I'm not an infra reviewer though.

ajsutton commented:
Oh turns out I am an infra reviewer. :) @jyellick Would you mind restoring the capturing of op-node logs and then I'll get this merged.

Thank you so much for your incredible patience - we've been really slow on a bunch of your PRs.

ajsutton commented:
@jyellick The world is apparently out to get us today - the devnet build failed because of a network error downloading node. Can you kick it to rerun and then I think we should be good to go. It won't let me rerun because it look like its running in your circle ci org.

jyellick commented:
Thanks @ajsutton, looks like the re-run fixed things, very happy to see this merge!
